app/src/main/java/com/zeapo/pwdstore/SearchableRepositoryViewModel.kt

```diff
@@ -0,0 +1,268 @@
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.mapLatest
+import kotlinx.coroutines.flow.toList
+import kotlinx.coroutines.yield
+
+private fun File.toPasswordItem(root: File) = if (isFile)
+    PasswordItem.newPassword(name, this, root)
+else
+    PasswordItem.newCategory(name, this, root)
+
+private fun PasswordItem.fuzzyMatch(filter: String): Int {
+    var i = 0
+    var j = 0
+    var score = 0
+    var bonus = 0
+    var bonusIncrement = 0
+
+    val toMatch = longName
+
+    while (i < filter.length && j < toMatch.length) {
+        when {
+            filter[i].isWhitespace() -> i++
+            filter[i].toLowerCase() == toMatch[j].toLowerCase() -> {
+                i++
+                bonusIncrement += 1
+                bonus += bonusIncrement
+                score += bonus
+            }
+            else -> {
+                bonus = 0
+                bonusIncrement = 0
+            }
+        }
+        j++
+    }
+    return if (i == filter.length) score else 0
+}
+
+private val CaseInsensitiveComparator = Collator.getInstance().apply {
+    strength = Collator.PRIMARY
+}
+
+private fun PasswordItem.Companion.makeComparator(
+    typeSortOrder: PasswordRepository.PasswordSortOrder,
+    directoryStructure: DirectoryStructure
+): Comparator<PasswordItem> {
+    return when (typeSortOrder) {
+        PasswordRepository.PasswordSortOrder.FOLDER_FIRST -> compareBy { it.type }
+        PasswordRepository.PasswordSortOrder.INDEPENDENT -> compareBy<PasswordItem>()
+        PasswordRepository.PasswordSortOrder.FILE_FIRST -> compareByDescending { it.type }
+    }
+        .then(compareBy(nullsLast(CaseInsensitiveComparator)) {
+            directoryStructure.getIdentifierFor(
+                it.file
+            )
+        })
+        .then(compareBy(CaseInsensitiveComparator) { directoryStructure.getUsernameFor(it.file) })
+}
+
+enum class FilterMode {
+    ListOnly,
+    StrictDomain,
+    Fuzzy
+}
+
+enum class SearchMode {
+    Recursive,
+    CurrentDirectoryOnly
+}
+
+private data class SearchAction(
+    val currentDir: File,
+    val filter: String = "",
+    val filterMode: FilterMode = FilterMode.ListOnly,
+    val searchMode: SearchMode = SearchMode.CurrentDirectoryOnly,
+    val listFilesOnly: Boolean = true
+)
+
+@ExperimentalCoroutinesApi
+@FlowPreview
+class SearchableRepositoryViewModel(application: Application) : AndroidViewModel(application) {
+    private val root = PasswordRepository.getRepositoryDirectory(application)
+    private val settings = PreferenceManager.getDefaultSharedPreferences(getApplication())
+    private val showHiddenDirs = settings.getBoolean("show_hidden_folders", false)
+    private val searchFromRoot = settings.getBoolean("search_from_root", false)
+    private val defaultSearchMode = if (settings.getBoolean(
+            "filter_recursively",
+            true
+        )
+    ) SearchMode.Recursive else SearchMode.CurrentDirectoryOnly
```

Comment:
indentation in this block looks pretty insane, would braces make this prettier?

Reply:
Narrator: They did.